### ISSUE TITLE
JIT 6502: Possible problem with re-entrant 6502 IRQs

### ISSUE THREAD

**Issue by hoglet67** [hoglet67/PiTubeDirect#123]
The following program will quickly hang when run with VSYNC events enabled:
```
10 ON ERROR RUN
20 I%=I%+1
30 PRINT I%
40 *WIBBLE
>*FX 14,4
>RUN
1
2
...
142
<hangs>
```
Here's a trace of the host side tube writes:
```
03.957469 : Mem Wr Watch hit at 0671 writing FEE7:FF  .
03.957497 : Mem Wr Watch hit at 0666 writing FEE3:00  .
03.957526 : Mem Wr Watch hit at 0666 writing FEE3:FE  .
03.957555 : Mem Wr Watch hit at 0666 writing FEE3:42  B
03.957589 : Mem Wr Watch hit at 0666 writing FEE3:61  a
03.957623 : Mem Wr Watch hit at 0666 writing FEE3:64  d
03.957657 : Mem Wr Watch hit at 0666 writing FEE3:20   
03.957691 : Mem Wr Watch hit at 0666 writing FEE3:63  c
03.957725 : Mem Wr Watch hit at 0666 writing FEE3:6F  o
03.957759 : Mem Wr Watch hit at 0666 writing FEE3:6D  m
03.957793 : Mem Wr Watch hit at 0666 writing FEE3:6D  m
03.957827 : Mem Wr Watch hit at 0666 writing FEE3:61  a
03.957861 : Mem Wr Watch hit at 0666 writing FEE3:6E  n
03.957895 : Mem Wr Watch hit at 0666 writing FEE3:64  d
03.957929 : Mem Wr Watch hit at 0666 writing FEE3:00  .
03.958231 : Mem Wr Watch hit at 0666 writing FEE3:A3  .
03.958247 : Mem Wr Watch hit at 0602 writing FEE3:09  .
03.958261 : Mem Wr Watch hit at 05E5 writing FEE3:00  .
03.958487 : Mem Wr Watch hit at 068F writing FEE1:80  .
03.958523 : Mem Wr Watch hit at 05E5 writing FEE3:00  .

03.983265 : Mem Wr Watch hit at 068F writing FEE1:00  .
03.983289 : Mem Wr Watch hit at 068F writing FEE1:10  .
03.983313 : Mem Wr Watch hit at 068F writing FEE1:00  .
03.983333 : Mem Wr Watch hit at 068F writing FEE1:04  .

04.023199 : Mem Wr Watch hit at 068F writing FEE1:00  .
04.023223 : Mem Wr Watch hit at 068F writing FEE1:0F  .
04.023247 : Mem Wr Watch hit at 068F writing FEE1:CB  .
04.023267 : Mem Wr Watch hit at 068F writing FEE1:04  .

04.049757 : Mem Wr Watch hit at 0671 writing FEE7:FF  .
04.049785 : Mem Wr Watch hit at 0666 writing FEE3:00  .
04.049814 : Mem Wr Watch hit at 0666 writing FEE3:FE  .
04.049843 : Mem Wr Watch hit at 0666 writing FEE3:42  B
04.049877 : Mem Wr Watch hit at 0666 writing FEE3:61  a
04.049911 : Mem Wr Watch hit at 0666 writing FEE3:64  d
04.049945 : Mem Wr Watch hit at 0666 writing FEE3:20   
04.049979 : Mem Wr Watch hit at 0666 writing FEE3:63  c
04.050013 : Mem Wr Watch hit at 0666 writing FEE3:6F  o
04.050047 : Mem Wr Watch hit at 0666 writing FEE3:6D  m
04.050081 : Mem Wr Watch hit at 0666 writing FEE3:6D  m
04.050115 : Mem Wr Watch hit at 0666 writing FEE3:61  a
04.050149 : Mem Wr Watch hit at 0666 writing FEE3:6E  n
04.050183 : Mem Wr Watch hit at 0666 writing FEE3:64  d
04.050217 : Mem Wr Watch hit at 0666 writing FEE3:00  .
04.050519 : Mem Wr Watch hit at 0666 writing FEE3:A3  .
04.050535 : Mem Wr Watch hit at 0602 writing FEE3:09  .
04.050549 : Mem Wr Watch hit at 05E5 writing FEE3:00  .
04.050775 : Mem Wr Watch hit at 068F writing FEE1:80  .
04.050811 : Mem Wr Watch hit at 05E5 writing FEE3:00  .

04.063135 : Mem Wr Watch hit at 068F writing FEE1:00  .
04.063159 : Mem Wr Watch hit at 068F writing FEE1:0E  .
04.063183 : Mem Wr Watch hit at 068F writing FEE1:01  .
04.063203 : Mem Wr Watch hit at 068F writing FEE1:04  .

04.103075 : Mem Wr Watch hit at 068F writing FEE1:00  .
04.103099 : Mem Wr Watch hit at 068F writing FEE1:0D  .
04.103123 : Mem Wr Watch hit at 068F writing FEE1:DD  .
04.103143 : Mem Wr Watch hit at 068F writing FEE1:04  .

04.142633 : Mem Wr Watch hit at 0671 writing FEE7:FF  .
04.142661 : Mem Wr Watch hit at 0666 writing FEE3:00  .
04.142690 : Mem Wr Watch hit at 0666 writing FEE3:FE  .
04.142719 : Mem Wr Watch hit at 0666 writing FEE3:42  B
04.142753 : Mem Wr Watch hit at 0666 writing FEE3:61  a
04.142787 : Mem Wr Watch hit at 0666 writing FEE3:64  d
04.142821 : Mem Wr Watch hit at 0666 writing FEE3:20   

04.143009 : Mem Wr Watch hit at 068F writing FEE1:00  .
04.143033 : Mem Wr Watch hit at 068F writing FEE1:0C  .
04.143057 : Mem Wr Watch hit at 068F writing FEE1:20   
04.143077 : Mem Wr Watch hit at 068F writing FEE1:04  .

04.143155 : Mem Wr Watch hit at 0666 writing FEE3:63  c
04.143189 : Mem Wr Watch hit at 0666 writing FEE3:6F  o   <<<<< host stops sending because R2 full

04.182943 : Mem Wr Watch hit at 068F writing FEE1:00  .
04.182967 : Mem Wr Watch hit at 068F writing FEE1:0B  .
04.182991 : Mem Wr Watch hit at 068F writing FEE1:6F  o
04.183011 : Mem Wr Watch hit at 068F writing FEE1:04  .

04.222881 : Mem Wr Watch hit at 068F writing FEE1:00  .
04.222905 : Mem Wr Watch hit at 068F writing FEE1:0A  .
04.222929 : Mem Wr Watch hit at 068F writing FEE1:6F  o
04.222949 : Mem Wr Watch hit at 068F writing FEE1:04  .
```
Looking at the client side error handler, it seems the consumption of the error message happens with IRQ enabled, and the client can and does receive event during this time:
```
.LFD3F
        LDA     R4DATA     ;; this FF that was written to  FEE7
        BPL     LFD65

        CLI
.LFD45
        BIT     R2STATUS
        BPL     LFD45

        LDA     R2DATA
        LDA     #$00   ;; this is writing the BRK to the error block
        STA     L0236
        TAY
        JSR     LF975

        STA     L0237
.LFD59
        INY
        JSR     LF975

        STA     L0237,Y
        BNE     LFD59

        JMP     L0236   ;; this executes the BRK

```
In the case of a "hang" the client apparantly stops in the middle of the loop copying the error message.

It carries on processing additional event IRQ messages.

It's a complex case to think about!

**Comment by hoglet67** [comment]
I'm fairly sure the problem is that our FIQ handler is not fully re-entrant, due to the way the fiq_stack variable is used:
```
fiq_handler_jit::
...
   push {r0-r2, regSP}
   ldr regSP, fiq_stack
...
  < execute 6502 interrupt handler with fiq enabled >
...
   str regSP, fiq_stack
   pop {r0-r2, regSP}
...
```
In the case of a re-entrant 6502 IRQ, this becomes unbalanced and at the start of the second IRQ you end up resetting the 6502 stack pointer back to the start of the FIQ stack.

Stepping back....

The original problem (which a seperate FIQ 6502 stack solves) is when you get:

````
Main program (running most 6502 code)
    --> ARM IRQ
        --> ARM FIQ (running 6502 interrupt code)
````
The FIQ handler can no longer access the 6502 state (i.e. regSP) from the main program, because this now only exists on the ARM IRQ stack.

I wonder if an alternative solution would be for the IRQ handler to store the 6502 state at a fixed address, together with a liveness flag.  The FIQ handler would then load the state from there. I tried to work through the finer details of this, and it got complicated very quickly!

**Comment by hoglet67** [comment]
Dominic,

Take a look at this for a possible fix: aff0aaf9

It basically does the following:
- moves back to a single 6502 stack (i.e. gets rid of FIQSTACK)
- updates the ARM IRQ handler to save a copy of regsp at a fixed address
- updates the ARM FIQ handler to re-load regsp from this fixed location if SPSR shows the saved mode as IRQ

This seems a simpler fix to the original problem of the ARM IRQ handler clobbering the 6502 state.

It works because ARM IRQs are guaranteed not to be re-entrant.

I've done a bit of testing and have still observed some very occasional crashes.

The ioload/iosave related changes (included in the above) helped, but haven't fixed things completely.

I think there is a tiny window of opportunity here:
```
_main_irq_handler:
        str     r9, _main_irq_r9_regsp
        sub     lr, lr, #4
        push    {r0, r1, r2, r3, ip, lr}
        bl      RPI_AuxMiniUartIRQHandler
        ldm     sp!, {r0, r1, r2, r3, ip, pc}^
```
Could a, FIQ happen after the ARM has entered IRQ mode, but before its saved r9?

If so, then it could be fixed by zeroing _main_irq_r9_regsp at the end to flag it as invalid, and checking for this case in the FIQ handler.

**Comment by hoglet67** [comment]
It's possible to run the test case for much longer now.
![capture10](https://user-images.githubusercontent.com/1951463/131247483-028e7c1e-6080-4658-924d-9a3d6f487705.png)

The limiting factor is the stack leak (of 17 words): 512KB / 17 words = 7710

I'm running this test on a Pi 4 with the debug build and with vdu=1 so there are ARM IRQ vsync / timer interrupts occurring.

**Comment by hoglet67** [comment]
Merged fix aff0aaf9 back into hognose-dev
